File: apps/go-ingester/pkg/logger/logger.go
package logger

import (
	"log/slog"
	"os"
	"strings"

	"go-ingester/internal/config"
)

// Init initializes a structured JSON slog logger from config and sets it as
// the default logger. Returns the configured logger for explicit passing.
func Init(cfg *config.Config) *slog.Logger {
	level := parseLevel(cfg.Core.LogLevel)

	h := slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{Level: level})
	log := slog.New(h).With(
		slog.String("service", "aurora-ingester"),
		slog.String("env", cfg.Core.Env),
	)
	slog.SetDefault(log)

	log.Info("Structured logger initialized")
	return log
}

func parseLevel(s string) slog.Level {
	switch strings.ToLower(s) {
	case "debug":
		return slog.LevelDebug
	case "warn", "warning":
		return slog.LevelWarn
	case "error":
		return slog.LevelError
	default:
		return slog.LevelInfo
	}
}

File: apps/rust-preprocessor/pkg/logger.rs
use tracing_subscriber::{fmt, EnvFilter};

pub fn init(log_level: &str, env: &str) {
    let filter = EnvFilter::try_from_default_env()
        .unwrap_or_else(|_| EnvFilter::new(log_level));

    fmt()
        .json()
        .with_env_filter(filter)
        .with_target(false)
        .with_current_span(false)
        .init();

    tracing::info!(
        service = "aurora-preprocessor",
        env = env,
        "Structured JSON logger initialized"
    );
}
